### PR TITLE
Qualcomm AI Engine Direct - [LLM QAT] Embedding + LM head W4 PCQ support

### PR DESCRIPTION
**Pull request by DannyYuyang-quic**
### Summary
 - With W4 PCQ applied to the embedding and LM head, we can further compress the embedding size.
W4 PCQ.
- Note: PCQ for embeddings is supported only in `QNN SDK` 2.48 or later.

### E2E script
``` bash
python examples/qualcomm/oss_scripts/llama/llama.py --build_folder build-android --device ${SERIAL_NUM} --soc_model {SOC_MODEL} --decoder_model smollm2_135m --model_mode hybrid --max_seq_len 1024 --prompt "What dose it mean to edit written content?" --qat --train_hf_dataset "HuggingFaceTB/smol-smoltalk" --train_hf_limit 8000 --calib_hf_dataset "HuggingFaceTB/smol-smoltalk" --calib_hf_limit 8000 --batch_size 2 --grad_accum_steps 4
```

### Test plan
``` bash
python backends/qualcomm/tests/test_qnn_delegate.py TestExampleLLMScript.test_static_llm_qat --device ${SERIAL_NUM} --soc_model ${SOC_MODEL} -r . -a . --build_folder build-android
```

### CODE REVIEW

**Comment by psiddh** [comment]
@DannyYuyang-quic You said embedding only , but this also drops the LM head to 4-bit iiuc. Did you mean to? Does accuracy still hold?

File: examples/qualcomm/oss_scripts/llama/static_llm_quant_recipe.py [hunk 1]
```diff
@@ -725,14 +721,7 @@ def __init__(self, verbose: bool = False):
             )
             .add_regex(
                 {r"tok_embeddings"},
-                QuantDtype.use_16a8w,
-                True,
-                act_observer=MovingAverageMinMaxObserver,
-                granularity=QuantGranularity.PER_TENSOR,
-            )
-            .add_regex(
```
Comment (psiddh) [inline examples/qualcomm/oss_scripts/llama/static_llm_quant_recipe.py:733]: This line was the only thing keeping the lm head at 8-bit. The PR removes it, isn't?

Comment (DannyYuyang-quic) [reply examples/qualcomm/oss_scripts/llama/static_llm_quant_recipe.py:733]: @psiddh 
Yes, this was intentional. The token embedding and LM head are closely related in the learned semantic space, as both represent tokens from the same vocabulary. This relationship is particularly strong in SmolLM2-135M, where the token embedding and LM head are originally tied and therefore initialized from the same learned vocabulary representation.

By making the token embedding and LM head learnable parameters in our QAT flow, they can adapt to quantization errors. Given their shared initialization and closely related roles, there is no strong a priori reason to quantize only the token embedding to W4 while keeping the LM head at W8. Applying W4 per-channel quantization to both was therefore intentional.

However, this also increases the number of trainable parameters, so the model may require more representative training data and additional training steps to converge.

I also ran text-generation tests, and the outputs remained coherent and semantically meaningful. Representative examples are attached below.

``` bash

PyTorchObserver {"prefill_token_per_sec":1500,"decode_token_per_sec":368.547,"prompt_tokens":18,"generated_tokens":307,"model_load_start_ms":1753617551343,"model_load_end_ms":1753617551737,"inference_start_ms":1753617551737,"inference_end_ms":1753617552582,"prompt_eval_end_ms":1753617551749,"first_token_ms":1753617551749,"aggregate_sampling_time_ms":95,"SCALING_FACTOR_UNITS_PER_SECOND":1000}
[INFO] [Qnn ExecuTorch]: Destroy Qnn backend
/data/local/tmp/yuyazhua/executorch/static_llm/outputs/outputs.txt: 1 file pulled. 1.8 MB/s (1513 bytes in 0.001s)
/data/local/tmp/yuyazhua/executorch/static_llm/outputs/inference_speed.txt: 1 file pulled. 0.0 MB/s (7 bytes in 0.001s)
[INFO 2026-07-19 17:20:45,293 llama.py:290] Device Inference Results[0]:
<|im_start|>user
What dose it mean to edit written content?<|im_end|>
<|im_start|>assistant

1. **Editing:** This is the process of making changes to a written text to improve its clarity, coherence, and overall effectiveness. It involves reviewing the text, identifying areas that need improvement, and revising them into a clearer and more polished version.

2. **Purpose:** The primary goal of editing is to refine the text to make it more engaging, readable, and effective. This can include simplifying sentence structure, correcting grammatical errors, and enhancing the tone and style of the text.

3. **Key Steps:**

1. **Read the text aloud:** This step is crucial in the editing process. It allows you to get a feel for the text's content, tone, and overall style.

2. **Review the text:** Carefully review the text to ensure that it is clear, concise, and free of errors.

3. **Editing:** Once you have reviewed the text, you can refine the text by revising it, adding or removing words, or reorganizing the text to improve its clarity and coherence.

4. **Final Step:** Once you have refined the text, you can start writing the final version of the text. This is often the final stage of the editing process, and the final version is the result of your efforts.

The goal of editing is to refine the text to make it more effective, engaging, and polished. This can be a challenging task, but with the right approach, you can refine your text to make it more engaging, readable, and effective.<|im_end|>
```

**Comment by DannyYuyang-quic** [comment]
@psiddh Hi, I have updated the PR title and description to explicitly include the LM-head change and avoid any misunderstanding about the scope of this PR.

**Comment by shewu-quic** [comment]
Hi @psiddh,
If you have any concerns about this PR, please feel free to share them with us.
Thanks! :)

**Review by psiddh — APPROVED** [review]
